lib/snippets/deploy-to-gke.rb

```diff
@@ -0,0 +1,137 @@
+
+require 'open3'
+require 'securerandom'
+require 'json'
+require 'pry'
+
+class GKEDeployment
+  class FatalDeploymentError < StandardError; end
+
+  def initialize(namespace:, gcloud_deployment:, gcloud_project:, environment:, current_sha:, template_folder: nil)
+    @namespace = namespace
+    @gcloud_deployment = gcloud_deployment
+    @gcloud_project = gcloud_project
+    @current_sha = current_sha
+    @id = current_sha + "-#{SecureRandom.hex(4)}"
+    @template_path = template_folder || "config/kubernetes/#{environment}/"
+    enforce_required_parmas
+  end
+
+  def run
+    authorize_gcloud
+    fetch_clusters.each do |cluster|
+      puts "debug0"
+      set_kubectl_cluster(cluster)
+      apply_all_templates
+    end
+  rescue FatalDeploymentError => e
+    print_error(e.message)
+    exit 1
+  end
+
+  def template_variables
+    {
+      'current_sha' => @current_sha,
+      'deployment_id' => @id,
+    }
+  end
+
+  private
+
+  def enforce_required_parmas
+    [@namespace, @gcloud_project, @gcloud_deployment, @current_sha].each do |required_param|
+      raise ArgumentError, "#{required_param} is required" unless required_param && !required_param.empty?
+    end
+  end
+
+  def apply_all_templates
+    unless Dir[File.expand_path("../#{@template_path}", __FILE__)]
+      raise FatalDeploymentError, "No templates found in #{@template_path}"
+    end
+
+
+    Dir[File.expand_path("../#{@template_path}", __FILE__)].each do |file|
+      if File.extname == '.yml'
+        run_command("kubectl apply -f #{file} --namespace=#{@namespace}")
+      elsif File.extname == 'yml.erb'
+        render_and_apply_template(render_template(file))
+      end
+    end
+  end
+
+  def render_and_apply_template(file_path)
+    erb_template = ERB.new(File.read(file_path))
+    erb_binding = binding
+    template_variables.each do |var_name, value|
+      erb_binding.local_variable_set(var_name, value)
+    end
+    content = erb_template.result(erb_binding)
+
+    f = Tempfile.new(['kube_template', '.yml'])
+    f.write(content)
+    f.close
+    run_command("kubectl apply -f #{f.path} --namespace#{@namespace}")
+  ensure
+    f.unlink if f
+  end
+
+  def authorize_gcloud
+    return if system('gcloud container clusters list')
+
+    st = run_command("gcloud auth activate-service-account --key-file $GCLOUD_CREDENTIALS_DIR/#{@gcloud_project}.json")
+    raise FatalDeploymentError, "Failed to set gcloud project #{@gcloud_project}" unless st
+  end
+
+  def set_kubectl_cluster(cluster)
+    cluster_name = cluster[0]
+    cluster_zone = cluster[1]
+    st = run_command("gcloud container clusters get-credentials #{cluster_name} --zone #{cluster_zone}")
+    raise FatalDeploymentError, "Failed to set cluster #{cluster_name}/#{cluster_zone}" unless st
+  end
+
+  def fetch_clusters
+    cmd = "gcloud deployment-manager deployments describe #{@gcloud_deployment} --format=json"
+    result = JSON.parse(`#{cmd}`)
```

Comment:
Review this line for a possible security issue: [Possible command injection](http://brakemanscanner.org/docs/warning_types/command_injection/).
in the following block of code:
```ruby
`#{"gcloud deployment-manager deployments describe #{@gcloud_deployment} --format=json"}`
```
this user-controlled input appears unsafe:
```ruby
@gcloud_deployment
```
Drop by #security on slack for help.